### PR TITLE
Implement SDL 0027 - Increase Specificity of Handler Events

### LINKED ISSUE #537: [SDL 0027] Increase Specificity of Handler Events
Proposal: [Increase Specificity of Handler Events](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0027-ios-specify-handlers.md)

> The purpose of this proposal is to make notification handlers more specific. Currently, the developer is required to know how to cast a notification parameter within the handler in order to properly use that notification. This proposal changes handlers to provide greater specificity on all RPCs that support them, as well as to separate out button handlers into two separate types. This will involve major version changes to remove old versions of the handlers that will no longer be supported.

Review: smartdevicelink/sdl_evolution#89

SDLC Steering Committee Decision:

> The Steering Committee has accepted this proposal, given the following revisions: the `SDLRPCButtonEventNotificationHandler` and `SDLRPCButtonPressNotificationHandler` would be combined into one handler instead, with the signature: `typedef void (^SDLButtonActionNotificationHandler)(__nullable SDLOnButtonPress *buttonPress, __nullable SDLOnButtonEvent *buttonEvent);`

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #537

This PR is **ready** for review.

### Risk
This PR makes **major** API changes.

### Summary
This PR implements SDL 0027 - Increase Specificity of Handler Events.

> The purpose of this proposal is to make notification handlers more specific. Currently, the developer is required to know how to cast a notification parameter within the handler in order to properly use that notification. This proposal changes handlers to provide greater specificity on all RPCs that support them, as well as to separate out button handlers into two separate types. This will involve major version changes to remove old versions of the handlers that will no longer be supported.

##### Breaking Changes
* Old event handler blocks have been updated with a new, specific signature for their respective types.

##### Enchancements
* Update button & command handlers to be specific.

### CODE REVIEW

**Comment by joeljfischer** [comment]
@asm09fsu updated

File: SmartDeviceLink/SDLNotificationConstants.h [hunk 1]
```diff
@@ -42,6 +39,26 @@ typedef void (^SDLRPCNotificationHandler)(__kindof SDLRPCNotification *notificat
...
+/**
+ A handler that may optionally be run when an SDLSubscribeButton or SDLSoftButton has a corresponding notification occur.
+
+ @warning This only works if you send the RPC using SDLManager.
+ @warning Only one of the two parameters will be set for each block call.
+
+ @param buttonPressNotification An SDLOnButtonPress object that corresponds to this particular button.
+ @param buttonEventNotification An SDLOnButtonEvent object that corresponds to this particular button.
+ */
+typedef void (^SDLRPCButtonNotificationHandler)(SDLOnButtonPress *_Nullable buttonPressNotification,  SDLOnButtonEvent *_Nullable buttonEventNotification);
```
Comment (asm09fsu) [inline SmartDeviceLink/SDLNotificationConstants.h]: can we shorten these to `buttonPress` and `buttonEvent`? We don't really care if it is a notification, really.

File: SmartDeviceLink/SDLNotificationConstants.h [hunk 2]
```diff
@@ -42,6 +39,26 @@ typedef void (^SDLRPCNotificationHandler)(__kindof SDLRPCNotification *notificat
...
+/**
+ A handler that may optionally be run when an SDLAddCommand has a corresponding notification occur.
+
+ @warning This only works if you send the RPC using SDLManager.
+
+ @param notification An SDLOnCommand object that corresponds to this particular SDLAddCommand.
+ */
+typedef void (^SDLRPCCommandNotificationHandler)(SDLOnCommand *notification);
```
Comment (asm09fsu) [inline SmartDeviceLink/SDLNotificationConstants.h]: Can we call this something else like `command` or `receivedCommand` or something? Notification is not a good name for this.

**Comment by joeljfischer** [comment]
@asm09fsu fixed

**Pull request merged**